01Matrix.js

```diff
@@ -0,0 +1,143 @@
+and the way that works is we're going to gather the location of all the 0's first and add them to the queue,
+that way we can process them together. During the iteration, we're going to BFS each of those 
+locations in all 4 directions and until we reach a 1. Once we reach a 1, we mark it with the current 
+traveled distance so far.
+
+Since we're using a BFS, we can tranverse in levels and mark the locations we have visited and skip the ones
+we have already visited. That way can keep track of the distance traveled so far and only update it
+if we're visiting a location for the first time and since it will be the first time, it will be
+the shorest path.
+*/
+
+/* 
+Question:
+Multi-sourced BFS solution, passes on LeetCode 
+
+Given an m x n binary matrix mat, return the distance of the nearest 0 for each cell.
+The distance between two adjacent cells is 1.
+*/
+
+const { Queue } = require('../../../utils/queue');
+
+//    0  1  2
+// 0 [0, 0, 0],
+// 1 [0, 1, 0],
+// 2 [1, 1, 1],
+
+const directions = [
+    [-1, 0],
+    [0, -1],
+    [0, 1],
+    [1, 0],
+];
+
+const getRowColString = (rowCol) => {
+    const { row, col } = rowCol;
+
+    return `${row}, ${col}`;
+};
+
+const getZeroFilledMatrix = (grid, fillWithValue) => {
+    const numRows = grid.length;
+    const numCols = grid[0].length;
+
+    const array = new Array(numRows);
+    for (let row = 0; row < numRows; row++) {
+        array[row] = new Array(numCols).fill(fillWithValue);
+    }
+
+    return array;
+};
+
+const getZeroLocations = (grid) => {
+    const locations = [];
+    const numRows = grid.length;
+    const numCols = grid[0].length;
+
+    for (let row = 0; row < numRows; row++) {
+        for (let col = 0; col < numCols; col++) {
+            if (grid[row][col] === 0) locations.push({ row, col });
+        }
+    }
+
+    return locations;
+};
+
+const isInBound = (row, col, grid) => {
+    const numRows = grid.length;
+    const numCols = grid[0].length;
+
+    const isRowInBound = row >= 0 && row < numRows;
+    const isColInBound = col >= 0 && col < numCols;
+
+    return isRowInBound && isColInBound;
+};
+
+const getChildren = (row, col, grid) => {
+    const children = [];
+
+    for (const direction of directions) {
+        const [rowDir, colDir] = direction;
+
+        const newRow = row + rowDir;
+        const newCol = col + colDir;
+
+        if (!isInBound(newRow, newCol, grid) || grid[newRow][newCol] === 0) continue;
+
+        children.push({ row: newRow, col: newCol });
+    }
+
+    return children;
+};
+
+const getDistanceFromZero = (grid) => {
+    const queue = new Queue();
+    const visited = new Set();
+
+    const zeroFilledGrid = getZeroFilledMatrix(grid, 0);
+    const zeroLocations = getZeroLocations(grid);
+
+    for (const zeroLocation of zeroLocations) {
+        visited.add(getRowColString(zeroLocation));
+        queue.enqueue({ ...zeroLocation, distanceFromZero: 0 });
```

Comment:
P: Why does your BFS start by processing the 0's if your goal is to mark distances for each of the 1's?

Reply:
Because if we're starting from the source, the path will always be 0. And since a BFS is increasing the path by 1 from each level-traversal (if it happens to be the first time visiting that location), we can keep track of the path so far until we reach our target and mark if with the path.

Reply:
Sort of. For your first paragraph, we can also do that if we started from the ones. The main issue is that the time complexity could get quite large if we continuously repeat BFS from each 1. Also, it would be a bit more complicated because when we find a 0, we'd have to remember which 1 to mark the distance for.

Reply:
Ahh, okay. I'm not very good at reading time complexity, yet, but I pulled this from the comments of a YouTube video. Starting from the 1's would have a time complexity of (m*n)^2 based off the number of 1's and the size of the grid. It could be wrong but could you explain what the time complexity would be BFS-ing from the 1's and then the 0's?

Reply:
It depends on how you implement it. 
Let r be the number of rows in the input matrix and let c be the number of columns.

There are two options for how you implement the algorithm:
1) iterate through the whole matrix. Whenever you find a 1, initiate a BFS until you find a 0. Mark the distance for the current 1 in the output matrix. Then continue iterating.
With this approach, we iterate through O(r*c) different nodes. And for each one, we initiate a BFS (we skip 0's but that doesn't matter for worst-case analysis). Our BFS could hit every node, so it runs in O(r*c) time. Thus, our overall complexity is O((r*c) * (r * c)) = O(r^2 * c^2).

2) do a multi-sourced BFS. So, we start by iterating over the whole matrix and adding every 1 to the queue. Then, we do a BFS and whenever we see a 0 we update the relevant nodes. Adding all the initial nodes is O(r*c) time and the BFS is also O(r*c) so the overall complexity is O(r*c + r*c) = O(r*c). However, the step of "updating the relevant nodes" is a bit complicated. It's not obvious to me how we'd do this, and I think it would take some explicit effort. We'd likely have to store extra info in the queue.

Because the first approach is slow and the second approach is complicated, it's easiest to BFS starting from the 0's.